man/dwdiff.1
.\" Generated by manscript from dwdiff.1.txt
.TH "DWDIFF" "1" "2013/03/10" "2.0.9" "Delimited word diff program"
.hw /usr/share/doc/dwdiff-2.0.9 http://os.ghalkes.nl/dwdiff.html
.SH NAME
dwdiff \- a delimited word diff program
.SH SYNOPSIS
\fBdwdiff\fR [\fIOPTIONS\fR] \fIOLD FILE\fR \fINEW FILE\fR
.br
\fBdwdiff\fR [\fIOPTIONS\fR] \fB\-\-diff\-input\fR [\fIDIFF FILE\fR]
.SH DESCRIPTION
\fBdwdiff\fR is a \fBdiff\fR program that operates at the word level instead of the line
level. It is different from \fBwdiff\fR in that it allows the user to specify what
should be considered whitespace, and in that it takes an optional list of
characters that should be considered delimiters. Delimiters are single
characters that are treated as if they are words, even when there is no
whitespace separating them from preceding words or delimiters. \fBdwdiff\fR is mostly
command-line compatible with \fBwdiff\fR. Only the \-\-autopager, \-\-terminal and
\-\-avoid\-wraps options are not supported.
.PP
The default output from \fBdwdiff\fR is the new text, with the deleted and inserted
parts annotated with markers. Command line options are available to change both
what is printed, and the markers.
.SH OPTIONS
\fBdwdiff\fR accepts the following options (Note that all strings will first be
escape expanded. All standard \\-escapes are supported, as well as \\u and \\U
Unicode escapes):
.PP
.TP
\fB\-h\fR, \fB\-\-help\fR
Display a short help message.
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version and copyright information.
.TP
\fB\-d\fR \fIdelimiters\fR, \fB\-\-delimiters\fR=\fIdelimiters\fR
Specify a list of characters to be used as delimiters.
.TP
\fB\-P\fR, \fB\-\-punctuation\fR
Use punctuation characters as delimiters. The exact set of punctuation
characters depends on the current locale.
.TP
\fB\-W\fR \fIwhitespace\fR, \fB\-\-whitespace\fR=\fIwhitespace\fR
Specify a list of characters to be used as whitespace.
.TP
\fB\-\-diff\-input\fR
Interpret the input as the output from \fBdiff\fR in the Unified Diff format
(usually produced by \fBdiff \-u\fR). In this case only one input file is
allowed. This option allows reformating diff output with \fBdwdiff\fR, and is
useful for example to post-process the output of \fBsvn diff\fR.
.TP
\fB\-1\fR, \fB\-\-no\-deleted\fR
Suppress printing of words deleted from the first file.
.TP
\fB\-2\fR, \fB\-\-no\-inserted\fR
Suppress printing of words inserted in the second file.
.TP
\fB\-3\fR, \fB\-\-no\-common\fR
Suppress printing of words common to both files.
.TP
\fB\-L\fR[\fIwidth\fR], \fB\-\-line\-numbers\fR[=\fIwidth\fR]
Show line numbers at the start of each line. The line numbers displayed are
the line number in the old file and the line number in the new file
respectively. The optional \fIwidth\fR argument is the minimum number of
positions per line number.
.TP
\fB\-C\fR\fInum\fR, \fB\-\-context\fR=\fInum\fR
Show \fInum\fR lines of context before and after each changes. A line
with only \-\- is printed between blocks of changes.
.TP
\fB\-s\fR, \fB\-\-statistics\fR
Print statistics when done. The numbers printed include the number of words
from in both files, the number of deleted words, the number of
inserted words, and the number of changed words. The number of changed words
is counted as the number of words that are removed from the first file, and
the number of words that replace them from the second file. All of these
numbers are also expressed as a percentage of the total number of words in the
file the words came from.
.TP
\fB\-i\fR, \fB\-\-ignore\-case\fR
Ignore differences in case when comparing words. This option is only
available if the diff program that is called provides it.
.TP
\fB\-I\fR, \fB\-\-ignore\-formatting\fR
Ignore differences in formatting of characters. This option switches to using
the Unicode compatibility decomposition instead of the canonical decomposition.
The compatibility decomposition discards formatting information. For example,
the ligature fi will be decomposed into two separate characters for the
purposes of comparison. However, also super- and subscript will be regarded
equal as well as different rotations of the same character.
.TP
\fB\-c\fR[\fIspec\fR], \fB\-\-color\fR[=\fIspec\fR]
Color mode. The optional \fIspec\fR can be used to customize the colors.
\fIspec\fR consists of [\fIdelete\fR],[\fIinsert\fR]. If either is
omited it will be set to its default color (bright red or bright green
respectively). Both parts of the \fIspec\fR consist of
[\fIforeground\fR][:\fIbackground\fR]. To obtain a list of permissible
color names, use the word ``list'' as \fIspec\fR. Alternatively, you can
specify any escape sequence to set attributes as a color by prepending
\fIe:\fR.
.IP
The standard markers for the begin and end of deleted and inserted text are
suppressed, but any markers specified on the command line will still be
printed.
.TP
\fB\-l\fR, \fB\-\-less\-mode\fR
As \-p but also overstrike deleted whitespace.
.TP
\fB\-p\fR, \fB\-\-printer\fR
Use overstriking with an underscore and bold text to emphasize changes. This
is implemented by first printing the underscore or a duplicate of the
character to be printed, followed by a backspace, followed by the character.
On regular terminals you won't see any effect. The \fBless\fR(1) command will
however show underlined and bold text.
.IP
The standard markers for the begin and end of deleted and inserted text are
suppressed, but any markers specified on the command line will still be
printed.
.TP
\fB\-m\fR\fInum\fR, \fB\-\-match\-context\fR=\fInum\fR
Use \fInum\fR words of context before and after words for matching. Words in
the old text will then only match words in the new text if words surrounding
them are also equal. This improves the output for \fBdwdiff\fR for large changes
with frequently occuring words. However, using context requires more disk
space and more processing time. The default value is 1. Set this option to 0
to revert to the pre 1.5 behavior.
.TP
\fB\-\-aggregate\-changes\fR
Allow multiple close changes to be treated as one change, if context words are
used (see \fB\-\-match\-context\fR). This option reduces the processing time as the
changes reported by the \fBdiff\fR program are not post-processed to give more
precise results.
.TP
\fB\-A\fR \fIalgorithm\fR, \fB\-\-algorithm\fR=\fIalgorithm\fR
Select the algorithm to be used for determining differences. There are three
possible values for \fIalgorithm\fR: \fIbest\fR, which tries to find the minimal set
of changes, \fInormal\fR, which trades some optimality for speed, and \fIfast\fR,
which assumes that the input is large and contains few changes. By default
the \fInormal\fR algorithm is used.
.TP
\fB\-S\fR[\fImarker\fR], \fB\-\-paragraph\-separator\fR[=\fImarker\fR]
Show insertion or deletion of blocks of lines with only whitespace characters.
A special marker is inserted into the output to indicate these blocks. The
default marker is \-\-.
.TP
\fB\-\-wdiff\-output\fR
Create \fBwdiff\fR compatible output. The \fBdwdiff\fR program uses a different output
algorithm, which provides a more intuitive output.
.TP
\fB\-w\fR \fIstring\fR, \fB\-\-start\-delete\fR=\fIstring\fR
Specify a string to mark begin of deleted text.
.TP
\fB\-x\fR \fIstring\fR, \fB\-\-stop\-delete\fR=\fIstring\fR
Specify a string to mark end of deleted text.
.TP
\fB\-y\fR \fIstring\fR, \fB\-\-start\-insert\fR=\fIstring\fR
Specify a string to mark begin of inserted text.
.TP
\fB\-z\fR \fIstring\fR, \fB\-\-stop\-insert\fR=\fIstring\fR
Specify a string to mark end of inserted text.
.TP
\fB\-R\fR, \fB\-\-repeat\-markers\fR
Repeat the begin and end markers at the start and end of line if a change
crosses a newline.
.PP
A single dash (\-) as a file can be used to denote standard input. Only one
file can be read from standard input. To stop \fBdwdiff\fR from interpreting
file names that start with a dash as options, one can specify a double dash
(\-\-) after which \fBdwdiff\fR will interpret any following arguments as files
to read.
.SH BUGS
If you think you have found a bug, please check that you are using the latest
version of \fBdwdiff\fR <http://os.ghalkes.nl/dwdiff.html>. When reporting
bugs, please include a minimal example that demonstrates the problem.
.SH AUTHOR
G.P. Halkes <dwdiff@ghalkes.nl>
.SH COPYRIGHT
Copyright \(co 2006\-2011 G.P. Halkes and others
.br
\fBdwdiff\fR is licensed under the GNU General Public License version 3.
.br
For more details on the license, see the file COPYING in the documentation
directory. On Un*x systems this is usually /usr/share/doc/dwdiff-2.0.9.
.SH SEE\ ALSO
\fBdwfilter\fR(1), \fBwdiff\fR(1), \fBdiff\fR(1)


Changelog
Version 2.0.9:
    Fix typo in configure script which prevents it from running.

Version 2.0.8:
    Bug fixes:
    - Fix build against glibc 2.16 and newer.

    Miscelaneous changes:
    - Added correct breaking for Unicode Regional Indicators.

Version 2.0.7:
    Fix incorrect release of 2.0.6, which omitted changes from the 2.0.5
    release.

Version 2.0.6
    Bug fixes:
    - Fix incorrect line numbering caused by optimized printing of differences
      preceeded in both files by whitespace containing newlines.

Version 2.0.5:
    Bug fixes:
    - fix a bug in the handling of context, which caused incorrect diffs to be
      printed. In some cases text was reported both as removed and as common.

Version 2.0.4:
    Bug fixes:
    - fix a memory allocation bug, which would be triggered by large values
      for the -m option.

Version 2.0.3:
    Updated translations for German (incomplete), French (incomplete) and
    Spanish.

Version 2.0.2:
    Bug fixes:
    - add support for DESTDIR in the Makefile install target.
    - fix build error on --without-unicode compiles.

Version 2.0.1:
    Bug fixes:
    - escape parsing for unknown escapes (like for example \;) would result
      in incorrect strings.
    - when the space character is not considered whitespace, dwdiff will no
      longer use it to make changes more readable.

    Internal changes:
    - the token file no longer uses a newline character as separator, but a nul
      character, making reparsing easier and the whole of dwdiff more robust.

Version 2.0:
    New features:
    - dwdiff is no longer a front-end for the diff program, but includes its
      own diff-algorithm implementation (taken from GNU diff). This makes
      dwdiff much faster, and ensures that results are the same across all
      platforms.

    Bug fixes:
    - when two additions are made with a single common token in between, dwdiff
      incorrectly reported the common token as replaced by the two additions
      and the common token.
    - when using the --diff-input option, in some cases dwdiff would report
      changes spanning multiple diff sections.

Version 1.9:
    New features:
    - an option for reading the output of diff (in Unified Diff) format was
      added (--diff-input) which can be used to reformat diffs for enhanced
      readability.

Version 1.8.2:
    New features:
    - when parsing options, \e is understood to mean the escape character

    Bug fixes:
    - the -p and -l options did not correctly handle UTF-8 characters.

Version 1.8.1:
    New features:
    - the context matching now decreases the matched context size exponentially,
      instead of immediately using the context-less mode if the changes found
      with context need refining. This results in more intuitive output when
      increasing the matching context size.

Version 1.8:
    New features:
    - added an option to repeat the begin and end markers at the start and end
      of line if a change crosses a newline

    Bug fixes:
    - newline characters in the delimiter set where handled incorrectly in
      UTF-8 mode

Version 1.7:
    New features:
    - added dwfilter program which reformats the old text such that it matches
      the new text to allow other programs such as visual diff tools to show
      the differences better. dwfilter can also be used to generate regular
      diffs to patch the old text by reformating the new text as the old text
      before executing diff [with caveats].

    Bug fixes:
    - \u and \U escapes were parsed incorrectly
    - dwdiff crashed when generating statistics for empty files

Version 1.6.1:
    Bug fixes:
    - the line clearing implemented for better background color handling caused
      (parts of) lines to disappear in the output with cr/nl line endings
    - the new paragraph separators were not handled correctly with the context
      matching feature, causing crashes

Version 1.6:
    New features:
    - the output for deleted text was changed such that the whitespace following
      the deletion is now the whitespace that followed the deleted text, rather
      than the whitespace that is now before the next word in the new file
    - the output for changed text was changed in the case that the whitespace
      before the old text contains a newline. In this case, depending on whether
      the whitespace before the new text contains a newline, the new file's
      whitespace is replaced by a single space or the new text is printed first
    - the --wdiff-output option was added to revert to the old output mode
    - the color specification now also allows specification of the background
      color and custom escape sequences
    - a new option to add markers for deleted or inserted blocks of lines
      containing only whitespace (paragraph separators)

    Bug fixes:
    - octal escapes are now properly parsed
    - icu-config was called directly rather than through $ICU_CONFIG in the
      configure script, which prevented proper operation of the --icu-config
      option
    - compilation on systems without nl_langinfo was fixed

Version 1.5.2
    New features:
    - better option parsing, which allows short options to be grouped. For
      example -iPc instead of -i -P -c

    Bug fixes:
    - compatibility fixes to configure script

Version 1.5.1:
    Bug fixes:
    - Statistics are now printed to stderr instead of stdout such that the only
      thing printed on stdout is the text

src/dwdiff.c
/* Copyright (C) 2006-2011 G.P. Halkes
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 3, as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>

#include "definitions.h"
#include "option.h"
#include "util.h"
#include "stream.h"
#include "unicode.h"
#include "dispatch.h"
#include "buffer.h"
#include "hashtable.h"

typedef enum {
	NONE,
	WHITESPACE,
	WORD
} MatchState;

int differences = 0;
Statistics statistics;
bool UTF8Mode;

/** Contains the (partial) word currently being read in. We only need one copy
	of this for all files, because all files are read sequentially. */
static CharBuffer currentWord;

CharBuffer whitespaceBuffer;
bool tokenWritten;

/** Contains the last read character. This is a global variable, because many
    routines use the same data and would require constant passing of either
    @a charData or a pointer to @a charData. Using a global-variable makes more
    sense in this case. */
CharData charData;

static void writeEndOfToken(InputFile *file) {
	ValueType wordValue;
	sputc(file->tokens->stream, 0);

	wordValue = getValueFromContext(&currentWord);
	tokenWritten = true;
	VECTOR_APPEND(file->diffTokens, wordValue);
	/* Reset current word */
	currentWord.used = 0;
}

/*===============================================================*/
/* Single character (SC) versions of the classification and storage routines.
   Descriptions can be found in the definition of the DispatchTable struct. */

bool getNextCharSC(Stream *file) {
	return (charData.singleChar = sgetc(file)) != EOF;
}

bool isWhitespaceSC(void) {
	return TEST_BIT(option.whitespace, charData.singleChar);
}

bool isDelimiterSC(void) {
	return TEST_BIT(option.delimiters, charData.singleChar);
}

void writeTokenCharSC(InputFile *file) {
	int diffChar = option.ignoreCase ? tolower(charData.singleChar) : charData.singleChar;

	VECTOR_APPEND(currentWord, diffChar);

	if (charData.singleChar == 0 || charData.singleChar == 1)
		filePutc(file->tokens->stream->data.file, 1);

	filePutc(file->tokens->stream->data.file, charData.singleChar);
}

void writeWhitespaceCharSC(InputFile *file) {
	/* Don't want to change interface (yet), so prevent warning. */
	(void) file;

	if (charData.singleChar == 0 || charData.singleChar == 1)
		VECTOR_APPEND(whitespaceBuffer, 1);

	VECTOR_APPEND(whitespaceBuffer, charData.singleChar);
}

void writeWhitespaceDelimiterSC(InputFile *file) {
	sputc(file->whitespace->stream, 0);
}

#ifdef USE_UNICODE
/*===============================================================*/
/* UTF-8 versions of the classification and storage routines.
   Descriptions can be found in the definition of the DispatchTable struct. */

bool getNextCharUTF8(Stream *file) {
	bool retval = getCluster(file, &charData.UTF8Char.original);
	if (retval)
		decomposeChar(&charData);
	return retval;
}

bool isWhitespaceUTF8(void) {
	if (option.whitespaceSet)
		return bsearch(&charData.UTF8Char.converted, option.whitespaceList.data, option.whitespaceList.used,
			sizeof(UTF16Buffer), (int (*)(const void *, const void *)) compareUTF16Buffer) != NULL;
	else
		return isUTF16Whitespace(&charData.UTF8Char.converted);
	return false;
}

bool isDelimiterUTF8(void) {
	return bsearch(&charData.UTF8Char.converted, option.delimiterList.data, option.delimiterList.used,
		sizeof(UTF16Buffer), (int (*)(const void *, const void *)) compareUTF16Buffer) != NULL ||
		(option.punctuationMask && isUTF16Punct(&charData.UTF8Char.converted));
}

void writeTokenCharUTF8(InputFile *file) {
	UChar32 highSurrogate = 0;
	UTF16Buffer *writeBuffer;
	size_t i;

	if (option.ignoreCase) {
		casefoldChar(&charData);
		writeBuffer = &charData.UTF8Char.casefolded;
	} else {
		writeBuffer = &charData.UTF8Char.converted;
	}

	for (i = 0; i < writeBuffer->used; i++) {
		char utf8char[4];
		size_t bytes;

		if ((bytes = filteredConvertToUTF8(writeBuffer->data[i], utf8char, &highSurrogate)) == 0)
			continue;
		VECTOR_ALLOCATE(currentWord, bytes);
		memcpy(currentWord.data + currentWord.used, &utf8char, bytes);
		currentWord.used += bytes;
	}

	/* Write the "original" characters. Note that high and low surrogates
	   and other invalid characters have been converted to REPLACEMENT
	   CHARACTER. */
	if (charData.UTF8Char.original.data[0] == 0 || charData.UTF8Char.original.data[0] == 1) {
		sputc(file->tokens->stream, 1);
		sputc(file->tokens->stream, charData.UTF8Char.original.data[0]);
	} else {
		for (i = 0; i < charData.UTF8Char.original.used; i++)
			putuc(file->tokens->stream, charData.UTF8Char.original.data[i]);
	}
}

void writeWhitespaceCharUTF8(InputFile *file) {
	UChar32 highSurrogate = 0;
	size_t i;

	/* Don't want to change interface (yet), so prevent warning. */
	(void) file;

	/* 0 and 1 are always considered to be a grapheme cluster on their own, and
	   are therefore always the only thing in the charData buffer if we
	   encouter them. Furthermore, we use 0 as line end, and 1 as escape
	   character in the temporary file. So we handle them separately here. */
	if (charData.UTF8Char.original.data[0] == 0 || charData.UTF8Char.original.data[0] == 1) {
		VECTOR_APPEND(whitespaceBuffer, 1);
		VECTOR_APPEND(whitespaceBuffer, charData.UTF8Char.original.data[0]);
		return;
	}

	for (i = 0; i < charData.UTF8Char.original.used; i++) {
		char utf8char[4];
		size_t bytes;

		if ((bytes = filteredConvertToUTF8(charData.UTF8Char.original.data[i], utf8char, &highSurrogate)) == 0)
			continue;

		VECTOR_ALLOCATE(whitespaceBuffer, bytes);

		memcpy(whitespaceBuffer.data + whitespaceBuffer.used, utf8char, bytes);
		whitespaceBuffer.used += bytes;
	}
}

void writeWhitespaceDelimiterUTF8(InputFile *file) {
	putuc(file->whitespace->stream, 0);
}

/*===============================================================*/
#endif

DEF_TABLE(SC)
ONLY_UNICODE(DEF_TABLE(UTF8))
DispatchTable *dispatch = &SCDispatch;

/** Handle the end of a whitespace sequence.
	@param file The @a InputFile from which the whitespace came.

	If the paragraph delimiter mode is selected, this will check whether such
	a delimiter should be written and break up the whitespace if necessary.
*/
void handleWhitespaceEnd(InputFile *file) {
	if (option.paraDelim) {
		size_t i, firstNewline = 0;
		bool firstNewlineFound = false;

		for (i = 0; i < whitespaceBuffer.used; i++) {
			if (whitespaceBuffer.data[i] != '\n')
				continue;
			if (!firstNewlineFound) {
				firstNewlineFound = true;
				firstNewline = i;
				continue;
			}
			break;
		}

		/* The whitespace preceeding any text must be treated differently, as a
		   newline there results in an empty line. This is different from other
		   whitespace where two newlines are required for an empty line. */
		if (firstNewlineFound && !tokenWritten) {
			/* Write everything upto but excluding the first newline */
			swrite(file->whitespace->stream, whitespaceBuffer.data, firstNewline);
			writeWhitespaceDelimiter(file);

			swrite(file->whitespace->stream, whitespaceBuffer.data + firstNewline, whitespaceBuffer.used - firstNewline);
			writeWhitespaceDelimiter(file);
			writeEndOfToken(file);
			whitespaceBuffer.used = 0;
			return;
		}

		if (i != whitespaceBuffer.used) {
			/* Write everything upto and including the first newline */
			swrite(file->whitespace->stream, whitespaceBuffer.data, firstNewline + 1);
			writeWhitespaceDelimiter(file);

			swrite(file->whitespace->stream, whitespaceBuffer.data + firstNewline + 1, whitespaceBuffer.used - (firstNewline + 1));
			writeWhitespaceDelimiter(file);
			writeEndOfToken(file);
			whitespaceBuffer.used = 0;
			return;
		}
		/* Fall through to default case */
	}

	swrite(file->whitespace->stream, whitespaceBuffer.data, whitespaceBuffer.used);
	writeWhitespaceDelimiter(file);
	whitespaceBuffer.used = 0;
}

/** Classify the character read in ::charData. */
int classifyChar(void) {
	/* Need to make sure we test delimiters first, because in UTF8Mode
	   we can't simply remove any overlapping delimiters from the
	   whitespace list. */
	return isDelimiter() ? CAT_DELIMITER : (isWhitespace() ? CAT_WHITESPACE : CAT_OTHER);
}

/** Read a file and separate whitespace from the rest.
    @param file The @a InputFile to read.
    @return The number of "words" in @a file.

    The separated parts of @a file are put into temporary files. The temporary
    files' information is stored in the @a InputFile structure.

    For runs in which the newline character is not included in the whitespace list,
    the newline character is transliterated into the first character of the
    whitespace list. Just before writing the output the characters are again
    transliterated to restore the original text.
*/
static int readFile(InputFile *file) {
	MatchState state = NONE;
	int wordCount = 0;
	int category;

	if (file->name != NULL && (file->input = newFileStream(fileOpen(file->name, FILE_READ))) == NULL)
		fatal(_("Can't open file %s: %s\n"), file->name, strerror(errno));

	if ((file->tokens = tempFile()) == NULL)
		fatal(_("Could not create temporary file: %s\n"), strerror(errno));

	VECTOR_INIT(file->diffTokens);

	if ((file->whitespace = tempFile()) == NULL)
		fatal(_("Could not create temporary file: %s\n"), strerror(errno));

	tokenWritten = false;

	while (getNextChar(file->input)) {
		category = classifyChar();
		switch (state) {
			case NONE:
				if (category == CAT_WHITESPACE) {
					writeWhitespaceChar(file);
					state = WHITESPACE;
					break;
				}
				handleWhitespaceEnd(file);
				writeTokenChar(file);
				if (category == CAT_DELIMITER) {
					writeEndOfToken(file);
					state = WHITESPACE;
				} else {
					state = WORD;
				}
				break;
			case WORD:
				if (category == CAT_WHITESPACE) {
					/* Found the end of a "word". Go to whitespace mode. */
					wordCount++;
					writeEndOfToken(file);
					writeWhitespaceChar(file);
					state = WHITESPACE;
				} else if (category == CAT_DELIMITER) {
					/* Found a delimiter. Finish the current word, add a zero length whitespace
					   to the whitespace file, add the delimiter as a word, and go into
					   whitespace mode. */
					wordCount += 2;
					writeEndOfToken(file);
					writeTokenChar(file);
					writeEndOfToken(file);
					handleWhitespaceEnd(file);
					state = WHITESPACE;
				} else {
					writeTokenChar(file);
				}
				break;
			case WHITESPACE:
				if (category == CAT_WHITESPACE) {
					writeWhitespaceChar(file);
				} else if (category == CAT_DELIMITER) {
					/* Found a delimiter. Finish the current whitespace, and add the delimiter
					   as a word. Then start new whitespace. */
					wordCount++;
					writeTokenChar(file);
					writeEndOfToken(file);
					handleWhitespaceEnd(file);
				} else {
					/* Found the start of a word. Finish the whitespace, and go into
					   word mode. */
					handleWhitespaceEnd(file);
					writeTokenChar(file);
					state = WORD;
				}
				break;
			default:
				PANIC();
		}
	}

	if (sferror(file->input))
		fatal(_("Error reading file %s: %s\n"), file->name, strerror(sgeterrno(file->input)));


	/* Make sure there is whitespace to end the output with. This may
	   be zero-length. */
	handleWhitespaceEnd(file);

	/* Make sure the word is terminated, or otherwise diff will add
	   extra output. */
	if (state == WORD) {
		wordCount++;
		writeEndOfToken(file);
	}
	/* Close the input, and make sure the output is in the filesystem.
	   Then rewind so we can start reading from the start. */
	sfclose(file->input);

	sfflush(file->whitespace->stream);
	if (sferror(file->whitespace->stream))
		fatal(_("Error writing to temporary file %s: %s\n"), file->name, strerror(sgeterrno(file->whitespace->stream)));
	srewind(file->whitespace->stream);

	sfflush(file->tokens->stream);
	if (sferror(file->tokens->stream))
		fatal(_("Error writing to temporary file %s: %s\n"), file->name, strerror(sgeterrno(file->tokens->stream)));
	srewind(file->tokens->stream);

	return wordCount;
}

/** Read the input files and perform the diff. */
static void prepareAndExecuteDiff(void) {
	statistics.oldTotal = readFile(&option.oldFile);
	statistics.newTotal = readFile(&option.newFile);
	baseHashMax = getHashMax();

	/* Whitespace buffer and currentWord won't be used after this. */
	VECTOR_FREE(currentWord);
	VECTOR_FREE(whitespaceBuffer);

	doDiff();
}

typedef enum {
	FIRST_HEADER,
	FIRST,
	OLD,
	NEW,
	COMMON,
	HEADER,
	LINE_COUNTS
} DiffInputMode;

/** Split the input, if it is the output from diff -u or similar. */
void splitDiffInput(void) {
	Stream *input;
	TempFile *oldFile, *newFile;
	DiffInputMode mode = FIRST_HEADER;

	/* when --diff-input is used option.oldFile.name
	will be NULL and option.oldFile.input will be a file stream
	to the diff.
	See option.c for details */
	if (option.oldFile.name == NULL) {
		input = option.oldFile.input;
	} else {
		if ((input = newFileStream(fileOpen(option.oldFile.name, FILE_READ))) == NULL)
			fatal(_("Can't open file %s: %s\n"), option.oldFile.name, strerror(errno));
	}

	oldFile = tempFile();
	newFile = tempFile();

	/* 
	Iterate over the characters of the input.
	We switch on the current "mode" (1st time through it'll be FIRST_HEADER).
	In each case block we first take the current character and either:
		* sputc to newFile->stream and/or oldFile->stream
		* putchar
	When the character is a newline we set the mode to FIRST.
	If the mode is FIRST we analize the character under the belief 
		that it will contain one of the initial lines from a diff
		which tells us what kind of line we're dealing with:
		NEW, OLD, or COMMON.  E.g. a line that has something added, 
		removed, or is common between the two files.

		When doing the sputc in this block all the special characters
		are replaced with a space. [It's not clear to me why we bother 
		with a sputc at all in this case. -masukomi]

	*/
	while (getNextCharSC(input)) { 
		// iterate over the characters of the input
		switch (mode) {
			case FIRST_HEADER: // this is the state mode is initialized in
				putchar(charData.singleChar); // putchar to STDOUT
				mode = charData.singleChar != '@' ? HEADER : LINE_COUNTS;
				break;
			case FIRST:
				// SOMETHING ADDED
				if (charData.singleChar == '+') {
					sputc(newFile->stream, ' ');
					mode = NEW;
				// SOMETHING REMOVED
				} else if (charData.singleChar == '-') {
					sputc(oldFile->stream, ' ');
					mode = OLD; 
				// LINE WITHOUT CHANGE
				} else if (charData.singleChar == ' ') {
					sputc(oldFile->stream, ' ');
					sputc(newFile->stream, ' ');
					mode = COMMON; 
				} else {
					closeTempFile(oldFile);
					closeTempFile(newFile);

					option.oldFile.name = oldFile->name;
					option.newFile.name = newFile->name;

					prepareAndExecuteDiff();

					resetTempFiles();
					oldFile = tempFile();
					newFile = tempFile();

					putchar(charData.singleChar);
					// "Second verse: same as the first."
					mode = charData.singleChar == '@' ? LINE_COUNTS : HEADER;
				}
				break;
			case OLD: // sputc to oldFile stream
				sputc(oldFile->stream, charData.singleChar);
				if (charData.singleChar == '\n')
					mode = FIRST;
				break;
			case NEW: // sputc to newFile stream
				sputc(newFile->stream, charData.singleChar);
				if (charData.singleChar == '\n')
					mode = FIRST;
				break;
			case COMMON: // sputc to both
				sputc(oldFile->stream, charData.singleChar);
				sputc(newFile->stream, charData.singleChar);
				if (charData.singleChar == '\n')
					mode = FIRST;
				break;

			case HEADER: // putchar to STDOUT
				putchar(charData.singleChar);
				if (charData.singleChar == '\n')
					mode = FIRST_HEADER;
				break;

			case LINE_COUNTS: // putchar to STDOUT
				putchar(charData.singleChar);
				if (charData.singleChar == '\n')
					mode = FIRST;
				break;
			default:
				PANIC();
		}
	}
	closeTempFile(oldFile);
	closeTempFile(newFile);

	option.oldFile.name = oldFile->name;
	option.newFile.name = newFile->name;

	prepareAndExecuteDiff();
}

/** Main. */
int main(int argc, char *argv[]) {
#if defined(USE_GETTEXT) || defined(USE_UNICODE)
	setlocale(LC_ALL, "");
#endif

#ifdef USE_GETTEXT
	bindtextdomain("dwdiff", LOCALEDIR);
	textdomain("dwdiff");
#endif

#ifdef USE_UNICODE
	/* Check whether the input is UTF-8 encoded. */
#ifdef USE_NL_LANGINFO
	if (strcmp(nl_langinfo(CODESET), "UTF-8") == 0)
		UTF8Mode = true;
#else
	{
		char *lc_ctype, *location;
		int i;

		if ((lc_ctype = setlocale(LC_CTYPE, NULL)) == NULL)
			goto end_utf8_check;
		lc_ctype = safe_strdup(lc_ctype);

		/* Use ASCII specific tolower function here, because it is not certain
		   that using tolower with the user locale will give correct results. */
		for (i = strlen(lc_ctype) - 1; i >= 0; i--)
			lc_ctype[i] = ASCIItolower(lc_ctype[i]);

		if ((location = strstr(lc_ctype, ".utf8")) != NULL) {
			if (location[5] == 0 || location[5] == '@')
				UTF8Mode = true;
		} else if ((location = strstr(lc_ctype, ".utf-8")) != NULL) {
			if (location[6] == 0 || location[6] == '@')
				UTF8Mode = true;
		}
	}
  end_utf8_check:
#endif // USE_NL_LANGINFO
#endif // USE_UNICODE

#ifdef USE_UNICODE
	if (UTF8Mode) {
		VECTOR_INIT_ALLOCATED(charData.UTF8Char.original);
		VECTOR_INIT_ALLOCATED(charData.UTF8Char.converted);
		VECTOR_INIT_ALLOCATED(charData.UTF8Char.casefolded);
		dispatch = &UTF8Dispatch;
	}
#endif
	VECTOR_INIT(currentWord);

	parseCmdLine(argc, argv);

	VECTOR_INIT(whitespaceBuffer);

	/* If we are reading the output from diff -u, then we need to first split
	   the input into two separate files. After that, we can use our normal
	   algorithm for determining the difference between two files. */
	if (option.diffInput)
		splitDiffInput();
	else
		prepareAndExecuteDiff();

	fflush(option.output);

	if (option.statistics) {
		int common = statistics.oldTotal - statistics.deleted - statistics.oldChanged;
		if (statistics.oldTotal == 0) {
			fprintf(stderr, _("old: 0 words\n"));
		} else {
			fprintf(stderr, _("old: %d words  %d %d%% common  %d %d%% deleted  %d %d%% changed\n"), statistics.oldTotal,
				common, (common * 100)/statistics.oldTotal,
				statistics.deleted, (statistics.deleted * 100) / statistics.oldTotal,
				statistics.oldChanged, (statistics.oldChanged * 100) / statistics.oldTotal);
		}
		common = statistics.newTotal - statistics.added - statistics.newChanged;
		if (statistics.newTotal == 0) {
			fprintf(stderr, _("new: 0 words\n"));
		} else {
			fprintf(stderr, _("new: %d words  %d %d%% common  %d %d%% inserted  %d %d%% changed\n"), statistics.newTotal,
				common, (common * 100)/statistics.newTotal,
				statistics.added, (statistics.added * 100) / statistics.newTotal,
				statistics.newChanged, (statistics.newChanged * 100) / statistics.newTotal);
		}
	}

#ifdef DEBUG_MEMORY
	free(option.oldFile.diffTokens.data);
	free(option.newFile.diffTokens.data);
	free(charData.UTF8Char.original.data);
	free(charData.UTF8Char.converted.data);
	free(charData.UTF8Char.casefolded.data);
	free(option.delColor);
	free(option.addColor);
#endif
	return differences;
}
